File: backend/app/models/reading.py
from sqlalchemy import Column, Integer, String, Float, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from datetime import datetime
from app.core.database import Base

class Reading(Base):
    __tablename__ = "readings"

    id = Column(Integer, primary_key=True, index=True)
    mission_id = Column(String, ForeignKey("missions.mission_id"), index=True)
    data_source = Column(String, default="UNKNOWN") # CSV, ESP32, DEMO
    
    timestamp = Column(DateTime, index=True, nullable=False)
    
    latitude = Column(Float, nullable=False)
    longitude = Column(Float, nullable=False)
    altitude = Column(Float, nullable=True)
    altitude_reference = Column(String, default="RELATIVE_HOME")
    pm1 = Column(Float, nullable=True)
    pm25 = Column(Float, nullable=True)
    pm10 = Column(Float, nullable=True)
    
    temperature = Column(Float, nullable=True)
    humidity = Column(Float, nullable=True)
    
    speed = Column(Float, nullable=True)
    heading = Column(Float, nullable=True)
    battery = Column(Integer, nullable=True)
    satellites = Column(Integer, nullable=True)
    gps_status = Column(String, nullable=True)
    signal_strength = Column(Float, nullable=True)
    
    aqi = Column(Integer, nullable=True)
    aqi_category = Column(String, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)

    mission = relationship("Mission", back_populates="readings")

File: backend/models/Reading.js
const mongoose = require('mongoose');

const readingSchema = new mongoose.Schema({
  mission_id: { type: String, index: true },
  data_source: { type: String, default: "UNKNOWN" },
  
  timestamp: { type: Date, required: true, index: true },
  
  latitude: { type: Number, required: true },
  longitude: { type: Number, required: true },
  altitude: { type: Number, default: null },
  altitude_reference: { type: String, default: "RELATIVE_HOME" },
  
  pm1: { type: Number, default: null },
  pm25: { type: Number, default: null },
  pm10: { type: Number, default: null },
  
  temperature: { type: Number, default: null },
  humidity: { type: Number, default: null },
  
  speed: { type: Number, default: null },
  heading: { type: Number, default: null },
  battery: { type: Number, default: null },
  satellites: { type: Number, default: null },
  gps_status: { type: String, default: null },
  signal_strength: { type: Number, default: null },
  
  aqi: { type: Number, default: null },
  aqi_category: { type: String, default: null },
  
  created_at: { type: Date, default: Date.now }
});

// Add virtual properties similar to Python
readingSchema.virtual('pm25_aqi').get(function() {
  const { calculateAQI } = require('../services/aqi');
  if (this.pm25 !== null) {
    const result = calculateAQI({ pm25: this.pm25 });
    return result ? result.aqi : null;
  }
  return null;
});

readingSchema.virtual('pm10_aqi').get(function() {
  const { calculateAQI } = require('../services/aqi');
  if (this.pm10 !== null) {
    const result = calculateAQI({ pm10: this.pm10 });
    return result ? result.aqi : null;
  }
  return null;
});

// Ensure virtual fields are serialized.
readingSchema.set('toJSON', {
  virtuals: true
});

module.exports = mongoose.model('Reading', readingSchema);
